packages/notifier/src/publish-kit.js

```diff
@@ -0,0 +1,198 @@
+ * @returns {AsyncIterator<T>}
+ */
+const makeLatestIterator = subscriber => {
+  let latestPublishCount;
+  return Far('LatestIterator', {
+    next: () => {
+      const pubList = E(subscriber).subscribeAfter(latestPublishCount);
+      // Without an onRejection, if pubList is a rejection, this will
+      // propagate the rejection as it should.
+      return E.when(pubList, ({ head, publishCount }) => {
+        latestPublishCount = publishCount;
+        return head;
+      });
+    },
+  });
+};
+
+/**
+ * Given a local or remote subscriber, returns a local AsyncIterable which
+ * provides "lossy" iterations of the underlying PublicationList.
+ * By "lossy", we mean that you may miss any published state if a more
+ * recent published state can be reported instead.
+ *
+ * If the underlying PublicationList is terminated, that terminal value will be
+ * reported losslessly.
+ *
+ * @template T
+ * @param {ERef<Subscriber<T>>} subscriber
+ * @returns {AsyncIterable<T>}
+ */
+export const subscribeLatest = subscriber => {
+  const iterable = Far('LatestIterable', {
+    [Symbol.asyncIterator]: () => makeLatestIterator(subscriber),
+  });
+  return iterable;
+};
+harden(subscribeLatest);
+
+/**
+ * Makes a `{ publisher, subscriber }` pair for doing efficient
+ * distributed pub/sub supporting both "each" and "latest" iteration
+ * of published values.
+ *
+ * @template T
+ * @returns {PublishKit<T>}
+ */
+export const makeEmptyPublishKit = () => {
+  /** @type {Promise<PublicationRecord<T>>} */
+  let tailP;
+  /** @type {undefined | ((value: ERef<PublicationRecord<T>>) => void)} */
+  let tailR;
+  ({ promise: tailP, resolve: tailR } = makePromiseKit());
+
+  let currentPublishCount = 0n;
+  let currentP = tailP;
+  const advanceCurrent = (done, value, rejection) => {
+    if (tailR === undefined) {
+      throw new Error('Cannot update state after termination.');
+    }
+
+    currentPublishCount += 1n;
+    currentP = tailP;
+    const resolveCurrent = tailR;
+
+    if (done) {
+      tailP = HandledPromise.reject(
+        new Error('Cannot read past end of iteration.'),
+      );
+      // Suppress unhandled rejection error.
+      tailP.catch(() => {});
+      tailR = undefined;
+    } else {
+      ({ promise: tailP, resolve: tailR } = makePromiseKit());
+    }
+
+    if (rejection) {
+      resolveCurrent(rejection);
+    } else {
+      resolveCurrent(
+        harden({
+          head: { value, done },
+          publishCount: currentPublishCount,
+          tail: tailP,
+        }),
+      );
+    }
+  };
+
+  /**
+   * @template T
+   * @type {Subscriber<T>}
+   */
+  const subscriber = Far('Subscriber', {
+    subscribeAfter: (publishCount = -1n) => {
+      assert.typeof(publishCount, 'bigint');
+      if (publishCount === currentPublishCount) {
+        return tailP;
+      } else if (publishCount < currentPublishCount) {
+        return currentP;
+      } else {
+        throw new Error('Invalid publish count');
+      }
```

Comment:
@Chris-Hibbert This rejection of future counts breaks some tests you added to the documentation repository a couple years ago (which also preclude fully updating the count type from number to bigint): https://github.com/Agoric/documentation/commit/99dd4abac726b9ecaf10479fefd6647cd272e13b#diff-68759f07a2341259c38a5cc8735ba2583781cb57dd7a71f7f78ae3b9ab6d7f79
> ```js
>   const updateCount = 1; // bogus
>   // #region getUpdateSince
>   const update = await notifier.getUpdateSince(updateCount);
> ```
> ```js
>     return E(notifier1)
>       .getUpdateSince(2)
>       .then(_update2 => {
>         t.is(52, state);
>         t.falsy(complete);
> 
>         updater.finish(97);
>         return E(notifier1)
>           .getUpdateSince(3)
>           .then(_update3 => {
>             t.is(97, state);
>             t.truthy(complete);
>             return true;
>           });
>       });
> ```

Is that behavior considered important to preserve?

Reply:
No, it's explicitly marked `bogus`, and was the setup for a code snippet that was included in a page that is now gone. If we had someone responsible for the docs, I'd ask them to make a pass through the code snippets and excise anything that isn't currently included in documentation.

The contract is that the argument to getUpdateSince() can be something that was previously published as updateCount. If it's anything but the current count, it's ignored.

Reply:
@Chris-Hibbert Great, please review https://github.com/Agoric/documentation/pull/681 at your earliest convenience.